src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -70,9 +71,10 @@ static int smart_amp_init(struct processing_module *mod)
 	}
 
 	/* Copy the pin formats */
-	bs = sizeof(sad->ipc4_cfg.input_pins) + sizeof(sad->ipc4_cfg.output_pin);
-	memcpy_s(sad->ipc4_cfg.input_pins, bs,
-		 base_cfg->base_cfg_ext.pin_formats, bs);
+	memcpy_s(sad->ipc4_cfg.input_pins, in_size,
+		 base_cfg->base_cfg_ext.pin_formats, in_size);
+	memcpy_s(&sad->ipc4_cfg.output_pin, out_size,
+		 &base_cfg->base_cfg_ext.pin_formats[in_size], out_size);
```

Comment:
is this correct? Shouldnt this be 
```memcpy_s(&sad->ipc4_cfg.output_pin, out_size,
		 &base_cfg->base_cfg_ext.pin_formats[SMART_AMP_NUM_IN_PINS], out_size);
```
or more simply:
``` 
sad->ipc4_cfg.output_pin = base_cfg->base_cfg_ext.pin_formats[SMART_AMP_NUM_IN_PINS];
```

Reply:
but the type of base_cfg_ext.pin_formats is ### **uint8_t**
struct ipc4_base_module_cfg_ext {
	/* specifies number of items in input_pins array. Maximum size is 8 */
	uint16_t nb_input_pins;
	/* specifies number of items in output_pins array. Maximum size is 8 */
	uint16_t nb_output_pins;
	uint8_t reserved[12];
	/* Specifies format of input pins followed by output pins.
	 * Pin format arrays may be non-continuous i.e. may contain pin #0 format
	 * followed by pin #2 format in case pin #1 will not be in use.
	 * FW assigned format of the pin based on pin_index, not on a position of
	 * the item in the array. Applies to both input and output pins.
	 */
	**uint8_t pin_formats[];**
} __attribute__((packed, aligned(4)));

Reply:
@andrula-song I'm not sure what you're trying to say. what you really want to do is this: 
```
diff --git a/src/samples/audio/smart_amp_test_ipc4.c b/src/samples/audio/smart_amp_test_ipc4.c
index e37fe3f74..e45c35d2e 100644
--- a/src/samples/audio/smart_amp_test_ipc4.c
+++ b/src/samples/audio/smart_amp_test_ipc4.c
@@ -47,6 +47,7 @@ static int smart_amp_init(struct processing_module *mod)
        int bs;
        int ret;
        const struct ipc4_base_module_extended_cfg *base_cfg = mod_data->cfg.init_data;
+       const size_t in_size, out_size;
 
        comp_dbg(dev, "smart_amp_init()");
        sad = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*sad));
@@ -70,9 +71,11 @@ static int smart_amp_init(struct processing_module *mod)
        }
 
        /* Copy the pin formats */
-       bs = sizeof(sad->ipc4_cfg.input_pins) + sizeof(sad->ipc4_cfg.output_pin);
-       memcpy_s(sad->ipc4_cfg.input_pins, bs,
-                base_cfg->base_cfg_ext.pin_formats, bs);
+       in_size = sizeof(struct ipc4_input_pin_format) * SMART_AMP_NUM_IN_PINS;
+       memcpy_s(sad->ipc4_cfg.input_pins, in_size, base_cfg->base_cfg_ext.pin_formats, in_size);
+       out_size = sizeof(struct ipc4_output_pin_format) * SMART_AMP_NUM_OUT_PINS;
+       memcpy_s(sad->ipc4_cfg.input_pins, in_size, base_cfg->base_cfg_ext.pin_formats + in_size,
+                in_size);
 
        mod->simple_copy = true;
 
```

Reply:
aligned, let's use sizeof(struct x) * num to calculate the size to make code more clear.

Reply:
```
	memcpy_s(sad->ipc4_cfg.input_pins, sizeof(sad->ipc4_cfg.input_pins),
		 base_cfg->base_cfg_ext.pin_formats, in_size);
	memcpy_s(&sad->ipc4_cfg.output_pin, sizeof(sad->ipc4_cfg.output_pin),
		 &base_cfg->base_cfg_ext.pin_formats[in_size], out_size);
```